### PR TITLE
Node-genesis-state

### PR DESCRIPTION
**Pull request by ScottyPoi**
Adds a method to `StateNetwork` class:  
`initGenesis()`

state clients may call this method to generate a sparse state state trie for accounts in its radius from the Ethereum mainnet genesis state.

The `Genesis State` for mainnet includes 8893 accounts with non-zero ETH balances.
This can be initialized from a 0.3 MB JSON file with a mapping of account address to balance.
A `Trie` constructed from these 8893 accounts will produce the genesis `StateRoot` (block 0).

From this in-memory genesis state trie, we can create a sparse trie with only the trie nodes needed to prove the accounts in the node's `radius`, and store those accounts in the `StateDB`.

This will enable a node, upon startup, to serve  `AccountTrieProofs` for its share of the set of original accounts.

While early Ethereum blocks rarely contain transactions, the account balance for each block's miner does increase, creating a new state root and new trie nodes.  Nodes in the StateNetwork can use the original set of known in-radius addresses to begin searching for content on the network.

### CODE REVIEW

**Comment by holgerd77** [comment]
Just to note that you can also use @ethereumjs/genesis for mainnet genesis to import if it makes sense for you here (https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/genesis), then you do not need to copy over the whole mainnet state here.

**Review by acolytec3 — COMMENTED** [review]
Looks great, except for `bind` :grimacing:

**Comment by acolytec3** [comment]
> Just to note that you can also use @ethereumjs/genesis for mainnet genesis to import if it makes sense for you here (https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/genesis), then you do not need to copy over the whole mainnet state here.

Hmm, I wondered about this.  We might want to avoid using `genesis` just because we don't want to load all of that in memory since the `genesis` package has this data stored in JS files instead of JSON so it gets loaded into memory on startup.

**Comment by ScottyPoi** [comment]
> > Just to note that you can also use @ethereumjs/genesis for mainnet genesis to import if it makes sense for you here (https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/genesis), then you do not need to copy over the whole mainnet state here.
> 
> Hmm, I wondered about this. We might want to avoid using `genesis` just because we don't want to load all of that in memory since the `genesis` package has this data stored in JS files instead of JSON so it gets loaded into memory on startup.

That was essentially my thinking.  starts with the minimum data,  does a bit of in-memory computation to build a trie.  then stores the nodes from the slice of trie it cares about in the DB, and lets the rest go.

File: packages/portalnetwork/src/networks/state/state.ts [hunk 1]
```diff
@@ -49,6 +50,8 @@ export class StateNetwork extends BaseNetwork {
     )
   }
 
+  public initGenesis = initStateTrie.bind(this)
```
[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/portalnetwork/src/networks/state/state.ts:53]: 1 line vs. 3 to do the same thing :shrug
```
  public async initGenesis() {
    await initStateTrie(this)
  }
  ```

File: packages/portalnetwork/src/networks/state/genesis.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  const db = (partialTrie.database().db as any)._database
+  const entries = [...db.entries()] as [string, Uint8Array][]
+  return entries
+}
+
+export async function genesisTrie(state: StateNetwork) {
+  return genesisDBEntries('0x' + state.enr.nodeId, state['nodeRadius'])
+}
+
+export async function initStateTrie(this: StateNetwork) {
```
[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/portalnetwork/src/networks/state/genesis.ts:82]: :+1

**Comment by holgerd77** [comment]
> > Just to note that you can also use @ethereumjs/genesis for mainnet genesis to import if it makes sense for you here (https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/genesis), then you do not need to copy over the whole mainnet state here.
> 
> Hmm, I wondered about this. We might want to avoid using `genesis` just because we don't want to load all of that in memory since the `genesis` package has this data stored in JS files instead of JSON so it gets loaded into memory on startup.

Hmm, if that’s the case I really wonder why you didn’t stop us or intervened when we’ve set up this package in the first place? 🤔

**Comment by acolytec3** [comment]
> > > Just to note that you can also use @ethereumjs/genesis for mainnet genesis to import if it makes sense for you here (https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/genesis), then you do not need to copy over the whole mainnet state here.
> > 
> > 
> > Hmm, I wondered about this. We might want to avoid using `genesis` just because we don't want to load all of that in memory since the `genesis` package has this data stored in JS files instead of JSON so it gets loaded into memory on startup.
> 
> Hmm, if that’s the case I really wonder why you didn’t stop us or intervened when we’ve set up this package in the first place? 🤔

I know we discussed it at some point.  It's all in JS files because of that issue we had with JSON imports when we switched to the hybrid CJS/ESM build and not being able to decide what to do about the import assertions that Node wanted us to add to all of the json imports.  I think we decided "for now" to just switch them to JS files.  The difference with Ultralight is that it's intended to be a light footprint (at least in theory, though we've never actually looked at it's memory footprint) and we should never have a need to hold the genesis state in memory.  We could still switch to the `genesis` package if there's a compelling reason to.

**Review by acolytec3 — APPROVED** [review]
Ship it

**Pull request merged**